pkg/ruler/api.go

```diff
@@ -268,3 +284,241 @@ func (r *Ruler) alerts(w http.ResponseWriter, req *http.Request) {
+	ErrNoGroupName = errors.New("a matching group name must be provided in the request")
+	// ErrNoRuleGroups signals the rule group requested does not exist
+	ErrNoRuleGroups = errors.New("no rule groups found")
+	// ErrBadRuleGroup is returned when the provided rule group can not be unmarshalled
+	ErrBadRuleGroup = errors.New("unable to decoded rule group")
+)
+
+// ValidateRuleGroup validates a rulegroup
+func ValidateRuleGroup(g rulefmt.RuleGroup) []error {
+	var errs []error
+	for i, r := range g.Rules {
+		for _, err := range r.Validate() {
+			var ruleName string
+			if r.Alert != "" {
+				ruleName = r.Alert
+			} else {
+				ruleName = r.Record
+			}
+			errs = append(errs, &rulefmt.Error{
+				Group:    g.Name,
+				Rule:     i,
+				RuleName: ruleName,
+				Err:      err,
+			})
+		}
+	}
+
+	return errs
+}
+
+func marshalAndSend(formatted interface{}, w http.ResponseWriter, logger log.Logger) {
+	d, err := yaml.Marshal(&formatted)
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshalling yaml rule groups", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/yaml")
+	if _, err := w.Write(d); err != nil {
+		level.Error(logger).Log("msg", "error writing yaml response", "err", err)
+		return
+	}
+}
+
+func respondAccepted(w http.ResponseWriter, logger log.Logger) {
+	b, err := json.Marshal(&response{
+		Status: "success",
+	})
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshaling json response", "err", err)
+		respondError(logger, w, "unable to marshal the requested data")
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+
+	// Return a status accepted because the rule has been stored and queued for polling, but is not currently active
+	w.WriteHeader(http.StatusAccepted)
+	if n, err := w.Write(b); err != nil {
+		level.Error(logger).Log("msg", "error writing response", "bytesWritten", n, "err", err)
+	}
+}
+
+// parseNamespace parses the namespace from the provided set of params, in this
+// api these params are derived from the url path
+func parseNamespace(params map[string]string) (string, error) {
+	namespace, exists := params["namespace"]
+	if !exists {
+		return "", ErrNoNamespace
+	}
+
+	namespace, err := url.PathUnescape(namespace)
+	if err != nil {
+		return "", err
+	}
+
+	return namespace, nil
+}
+
+// parseGroupName parses the group name from the provided set of params, in this
+// api these params are derived from the url path
+func parseGroupName(params map[string]string) (string, error) {
+	groupName, exists := params["groupName"]
+	if !exists {
+		return "", ErrNoGroupName
+	}
+
+	groupName, err := url.PathUnescape(groupName)
+	if err != nil {
+		return "", err
+	}
+
+	return groupName, nil
+}
+
+func parseRequest(req *http.Request, requireNamespace, requireGroup bool) (string, string, string, error) {
+	id, err := user.ExtractOrgID(req.Context())
```

Comment:
`orgId` or `userId`

Reply:
Technically we use `x-orgid-scope` across are stack. Weave has support for users but it's not used in cortex. In general userID and orgID are the same across are stack. I'll add an issue to make the nomenclature consistent.

Reply:
I meant to use one of those instead of just `id`, that seems little too short here.

Reply:
Of course, I decided to use userID. Just haven’t pushed the to the branch yet 😉